x-pack/plugins/lists/public/exceptions/components/builder/exception_items_renderer.tsx

```diff
@@ -394,12 +396,36 @@ export const ExceptionBuilderComponent = ({
     }
   }, [exceptions, handleAddNewExceptionItem]);
 
+  /**
+   * This component relies on the "exceptionListItems" to pre-fill its entries,
+   *  but any subsequent updates to the entries are not reflected back to
+   * the "exceptionListItems". To ensure correct behavior, we need to only
+   * fill the entries from the "exceptionListItems" during initialization.
+   *
+   * In the initialization phase, if there are "exceptionListItems" with
+   * pre-filled entries, the exceptions array will be empty. However,
+   * there are cases where the "exceptionListItems" may not be sent
+   * correctly during initialization, leading to the exceptions
+   * array being filled with empty entries. Therefore, we need to
+   *  check if the exception is correctly populated with a valid
+   * "field" when the "exceptionListItems" has entries. that's why
+   * "exceptionsEntriesPopulated" is used
+   *
+   * It's important to differentiate this case from when the user
+   * deletes all the entries and the "exceptionListItems" has pre-filled values.
+   * that's why "allEntriesDeleted" is used
+   *
+   *  deepEqual(exceptionListItems, exceptions) to handle the exceptionListItems in
+   * the EventFiltersFlyout
+   */
   useEffect(() => {
-    if (exceptionListItems.length > 0) {
+    if (!exceptionListItems.length || deepEqual(exceptionListItems, exceptions)) return;
+    const exceptionsEntriesPopulated = exceptions.some((exception) =>
+      exception.entries.some((entry) => entry.field)
+    );
+    if (!exceptionsEntriesPopulated && !areAllEntriesDeleted)
       setUpdateExceptions(exceptionListItems);
-    }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+  }, [areAllEntriesDeleted, exceptionListItems, exceptions, setUpdateExceptions]);
```

Comment:
I'd avoid such a custom logic which needs a long explanation in comment, if possible

---

if I understand correctly, the problem is that the content of `exceptionListItems` comes too late, only after first render. if we need to solve this issue _inside_ the `ExceptionBuilderComponent`, I suggest to use refs, too, but in a more general way, something like this:

```ts
  const hasInitializedExceptionsRef = useRef<boolean>(false);

  useEffect(() => {
    if (!hasInitializedExceptionsRef.current && exceptionListItems.length > 0) {
      setUpdateExceptions(exceptionListItems);
      hasInitializedExceptionsRef.current = true;
    }
  }, [exceptionListItems, setUpdateExceptions]);
```
we can go even further, and modify the type of `exceptionListItems` so it is either `null` or `array`, so we know that if it's `null`, it's uninitialized.

---

another way would be to solve it on the outside of the component, i.e. we say that the contract of `ExceptionBuilderComponent` declares that it uses the content of `exceptionListItems` only in the first render. this requirement could be met by rendering the component conditionally, only _after_ we got the content, e.g. like this:

```ts
x-pack/plugins/security_solution/public/detection_engine/rule_exceptions/components/flyout_components/item_conditions/index.tsx:

      {exceptionListItems.length > 0 &&
        getExceptionBuilderComponentLazy({
          [...props come here...]
        })}
```

what do you think?

Reply:
@gergoabraham definitely agree with the points you made. I'd chatted with @WafaaNasr and she'd noted there should be a more thorough fix, but given we're in post-feature freeze zone, we're trying to stick to minimal changes that address the bug and follow up with a fix to address the underlying issue 8.10. 

@WafaaNasr do we know if there's any existing ticket to describe some of the issues with this component? Maybe we should add the 8.10 candidate tag to it, and if not, create one to track for 8.10.

Reply:
Thank you, @gergoabraham! Regarding the flashing issue, it occurred when I deleted all the pre-filled fields on a Rule Exception and added new fields along with a name. I tried to capture it on video, but unfortunately, I couldn't catch it again.

@yctercero, I thought that would be part of this [ticket](https://github.com/elastic/kibana/issues/154994), as part of the design, let me know if not I will create a new one.

Reply:
thanks for the explanation, totally understandable @yctercero @WafaaNasr! it's great that you already have a refactor ticket 👏